src/app/layers/qgsapplayerhandling.cpp

```diff
@@ -394,19 +394,23 @@ QList< QgsMapLayer * > QgsAppLayerHandling::addOgrVectorLayers( const QStringLis
   QgsProject::instance()->addMapLayers( layersToAdd );
   for ( QgsMapLayer *l : std::as_const( layersToAdd ) )
   {
-    if ( !encoding.isEmpty() )
-    {
-      if ( QgsVectorLayer *vl = qobject_cast< QgsVectorLayer * >( l ) )
-        vl->setProviderEncoding( encoding );
-    }
-
     QgisApp::instance()->askUserForDatumTransform( l->crs(), QgsProject::instance()->crs(), l );
     QgsAppLayerHandling::postProcessAddedLayer( l );
   }
   QgisApp::instance()->activateDeactivateLayerRelatedActions( QgisApp::instance()->activeLayer() );
 
   ok = true;
   addedLayers.append( layersToAdd );
+
+  for ( QgsMapLayer *l : std::as_const( addedLayers ) )
```

Comment:
Couldn't we check here if `l` is in `layersToAdd` and if so, do those steps https://github.com/qgis/QGIS/pull/52056/files#diff-8a3aa0d4bf3abb880d91f75b19afb6b8b5f91e0af0bc20081f1e1565253bc6f5R397-R398 to avoid to make two loops. But I guess it's a small list, so no big effort with that.

Reply:
Honestly I'm reticent to make too many changes to this logic -- these functions have a lot of non-obvious side effects and I'd also be concerned about making the code too unreadable (the difference between "layersToAdd" and "addedLayers" is quite hard to understand! :upside_down_face: )

Reply:
Fair enough.